numpy/core/src/multiarray/array_coercion.c

```diff
@@ -0,0 +1,1437 @@
+                return -1;
+            }
+            Py_XDECREF(item_DType);
+            Py_DECREF(elem);
+            PyArray_ITER_NEXT(iter);
+        }
+        Py_DECREF(iter);
+    }
+    else if (DType != NULL && NPY_UNLIKELY(DType->type_num == NPY_DATETIME) &&
+                PyArray_ISSTRING(arr)) {
+        /*
+         * TODO: This branch should be deprecated IMO, the workaround is
+         *       to cast to the object to a string array. Although a specific
+         *       function (if there is even any need) would be better.
+         *       This is value based casting!
+         * Unless of course we actually want to support this kind of thing
+         * in general (not just for object dtype)...
+         */
+        PyArray_DatetimeMetaData meta;
+        meta.base = NPY_FR_GENERIC;
+        meta.num = 1;
+
+        if (find_string_array_datetime64_type(arr, &meta) < 0) {
+            return -1;
+        }
+        else {
+            *out_descr = create_datetime_dtype(NPY_DATETIME, &meta);
+            if (*out_descr == NULL) {
+                return -1;
+            }
+        }
+    }
+    else {
+        /*
+         * If this is not an object array figure out the dtype cast,
+         * or simply use the returned DType.
+         */
+        *out_descr = cast_descriptor_to_fixed_dtype(
+                     PyArray_DESCR(arr), DType);
+        if (*out_descr == NULL) {
+            return -1;
+        }
+    }
+    return 0;
+}
+
+/**
+ * Given a dtype or DType object, find the correct descriptor to cast the
+ * array to.
+ *
+ * This function is identical to normal casting using only the dtype, however,
+ * it supports inspecting the elements when the array has object dtype
+ * (and the given datatype describes a parametric DType class).
+ *
+ * @param arr
+ * @param dtype A dtype instance or class.
+ * @return A concrete dtype instance or NULL
+ */
+NPY_NO_EXPORT PyArray_Descr *
+PyArray_AdaptDescriptorToArray(PyArrayObject *arr, PyObject *dtype)
+{
+    /* If the requested dtype is flexible, adapt it */
+    PyArray_Descr *new_dtype;
+    PyArray_DTypeMeta *new_DType;
+    int res;
+
+    res= PyArray_ExtractDTypeAndDescriptor((PyObject *)dtype,
+            &new_dtype, &new_DType);
+    if (res < 0) {
+        return NULL;
+    }
+    if (new_dtype == NULL) {
+        res = find_descriptor_from_array(arr, new_DType, &new_dtype);
+        if (res < 0) {
+            Py_DECREF(new_DType);
+            return NULL;
+        }
+        if (new_dtype == NULL) {
+            /* This is an object array but contained no elements, use default */
+            new_dtype = new_DType->default_descr(new_DType);
+        }
+    }
+    Py_DECREF(new_DType);
+    return new_dtype;
+}
+
+
+NPY_NO_EXPORT int
+PyArray_DiscoverDTypeAndShape_Recursive(
+        PyObject *obj, int curr_dims, int max_dims, PyArray_Descr**out_descr,
+        npy_intp out_shape[NPY_MAXDIMS],
+        coercion_cache_obj ***coercion_cache_tail_ptr,
```

Comment:
I need to take a deeper look here, but I couldn't understand the purpose of coercion_cache. Having some comments about coercion cache here would be great!

Reply:
Ah OK, the cache is consumed by `PyArray_AssignFromCache`, I should probably just reference that function, since it is the only consumer.  There are few cases were we do not actually need the cache, that could be an improvement to allow passing in `NULL` for the cache to indicate not filling it up.  It should be very simple (also later), will mean a bit of changes to correct for refcounting.

Reply:
Added a comment to the function with some details (and expanded the details in `PyArray_DiscoverDTypeAndShape`) admittedly, you probably have to read *both*, let me know if you have thoughts on more explanations!